src/test/java/at/ac/tuwien/kr/alpha/grounder/NaiveGrounderTest.java

```diff
@@ -97,14 +112,228 @@ public void groundConstraintAlreadyGround() {
+	}
+
+	private void testDeadEnd(RuleGroundingOrder groundingOrderP1, RuleGroundingOrder groundingOrderQ1, boolean expectNoGoods) {
+		Program program = PARSER.parse("p1(1). q1(1). "
+				+ "x :- p1(X), p2(X), q1(Y), q2(Y). "
+				+ "p2(X) :- something(X). "
+				+ "q2(X) :- something(X). ");
+
+		AtomStore atomStore = new AtomStoreImpl();
+		NaiveGrounder grounder = (NaiveGrounder) GrounderFactory.getInstance("naive", program, atomStore, p -> true, GrounderHeuristicsConfiguration.lax(), true);
+
+		NonGroundRule nonGroundRule = grounder.getNonGroundRule(0);
+		final Literal litP1X = literal("p1", "X");
+		final Literal litQ1Y = literal("q1", "Y");
+		nonGroundRule.groundingOrder.groundingOrders.put(litP1X, groundingOrderP1);
+		nonGroundRule.groundingOrder.groundingOrders.put(litQ1Y, groundingOrderQ1);
+
+		grounder.bootstrap();
+		TrailAssignment currentAssignment = new TrailAssignment(atomStore);
+		final Substitution substP1X1 = Substitution.unify(litP1X, new Instance(ConstantTerm.getInstance(1)), new Substitution());
+		final Substitution substQ1Y1 = Substitution.unify(litQ1Y, new Instance(ConstantTerm.getInstance(1)), new Substitution());
+		final NaiveGrounder.BindingResult bindingResultP1 = grounder.getGroundInstantiations(nonGroundRule, groundingOrderP1, substP1X1, currentAssignment);
+		final NaiveGrounder.BindingResult bindingResultQ1 = grounder.getGroundInstantiations(nonGroundRule, groundingOrderQ1, substQ1Y1, currentAssignment);
+
+		assertEquals(expectNoGoods, bindingResultP1.size() > 0);
+		assertEquals(expectNoGoods, bindingResultQ1.size() > 0);
+	}
+
+	@Test
+	public void testGroundingOfRuleSwitchedOffByFalsePositiveBody() {
+		Program program = PARSER.parse("a(1). "
+				+ "c(X) :- a(X), b(X). "
+				+ "b(X) :- something(X). ");
+		testIfGrounderGroundsRule(program, 0, literal("a", "X"), 1, ThriceTruth.FALSE, false);
+	}
+
+	@Test
+	public void testGroundingOfRuleNotSwitchedOffByTruePositiveBody() {
+		Program program = PARSER.parse("a(1). "
+				+ "c(X) :- a(X), b(X). "
+				+ "b(X) :- something(X). ");
+		testIfGrounderGroundsRule(program, 0, literal("a", "X"), 1, ThriceTruth.TRUE, true);
+	}
+
+	@Test
+	@Ignore("Currently, rule grounding is not switched off by a true negative body atom")
+	public void testGroundingOfRuleSwitchedOffByTrueNegativeBody() {
+		Program program = PARSER.parse("a(1). "
+				+ "c(X) :- a(X), not b(X). "
+				+ "b(X) :- something(X). ");
+		testIfGrounderGroundsRule(program, 0, literal("a", "X"), 1, ThriceTruth.TRUE, false);
+	}
+
+	@Test
+	public void testGroundingOfRuleNotSwitchedOffByFalseNegativeBody() {
+		Program program = PARSER.parse("a(1). "
+				+ "c(X) :- a(X), not b(X). "
+				+ "b(X) :- something(X). ");
+
+		testIfGrounderGroundsRule(program, 0, literal("a", "X"), 1, ThriceTruth.FALSE, true);
+	}
+
+	private void testIfGrounderGroundsRule(Program program, int ruleID, Literal startingLiteral, int startingInstance, ThriceTruth bTruth, boolean expectNoGoods) {
```

Comment:
It would be good if those private helpers were a bit documented.

Reply:
Done.